### ISSUE TITLE
Disallow if-in

### ISSUE THREAD

**Issue by Krinkle** [wikimedia/eslint-config-wikimedia#67]
It is almost always a mistake to use the `in` operator to check if a key exists. The reason being that, unlike for loops, the `in` operator in a conditional will not skip native prototype methods that would otherwise be excluded (enumerable false) from loops.

For example, we recommend the [guard-for-in](https://eslint.org/docs/rules/guard-for-in) rule because we don't support environments that alter the `Object.prototype` in ways that introduces enumerable properties. This is why it is safe and recommended in our code to do something like this:

```js
// This code is fine
var obj = {};
var key;
// ...
for (key in obj) {
 // ...
}
```

However, while for-in is fine, if-in can be dangerous and should be flagged by ESLint because all non-enumerable properties from the `Object.prototype` in browsers (such as "constructor", "toString", "valueOf", "watch", and a few more) will return true if explicitly compared.

```js
// This code is UNSAFE
var obj = {};
if (something in obj) {
  obj[something]++;
  // ^
  // Will break if something = "constructor", because
  // it would get value NaN (= nativeFunction + 1)
} else {
  obj[something] = 1;
}
```

Depending on the intention, a typical solution should involve one of these:

* using `Object.create(null)`, this is preferred. – You have control over the object.
* or using `hasOwnProperty`. – Foreign object, ignore inherited properties.
* or using a positive `typeof` check. – Foreign object, allow inherited properties with sanity check.

```js
// This code is fine (1)
var obj = Object.create(null); // <! changed
if (something in obj) {
  obj[something]++;
} else {
  obj[something] = 1;
}
```
```js
// This code is fine (2)
var obj = {};
var hasOwn = Object.hasOwnProperty; // <! added
if (hasOwn.call(obj, something)) { // <! changed
  obj[something]++;
} else {
  obj[something] = 1;
}
```
```js
// This code is fine (3)
var obj = {};
if (typeof obj[something] === 'number') { // <! changed
  obj[something]++;
} else {
  obj[something] = 1;
}
```

**Comment by jdforrester** [comment]
Seems sensible. Worth flagging on https://www.mediawiki.org/wiki/Manual_talk:Coding_conventions/JavaScript ?

**Comment by Krinkle** [comment]
> Worth flagging on https://www.mediawiki.org/wiki/Manual_talk:Coding_conventions/JavaScript ?

Yes.

As for ESLint, there exists no rule for this currently, but @platinumazure pointed out to me that the [no-restricted-syntax](https://eslint.org/docs/rules/no-restricted-syntax) rule can be used for this.

For example, selector `"IfStatement > BinaryExpression[operator='in'].test"` should catch the most common cases. Or the following, so that it catches negated forms as well: `":matches(IfStatement > BinaryExpression[operator='in'].test, IfStatement > .test BinaryExpression[operator='in'])"`.

Samples:

```js
// Caught:
if ("bar" in foo) {
}

// Caught:
if (!("baz" in foo)) {
}

// Not caught:
var x = "bat" in foo;
```

```json
{
    "rules": {
        "no-restricted-syntax": [
            "error",
            {
                "selector": ":matches(IfStatement > BinaryExpression[operator='in'].test, IfStatement > .test BinaryExpression[operator='in'])",
                "message": "Binary 'in' expression not allowed in 'if' statement.."
            }
        ]
    }
}
```

However, as mentioned in the original issue description, use of `if-in` is fine when using it against objects without inheritance, such as those created with `Object.create(null);` which is a fairly common and sensible idiom. The idiom is currently mostly used in our Node.js code, not so much in browser code. But that's only the case because most of our browser code still smells of "ES3" compatibility. It would be a fine pattern to adopt in browser code now. (Note, Object.create is ES5, not ES6+, so this is been widely usable for many years now and totally safe in MediaWiki JS as well).

On the other hand, for cases where the object cannot be changed, or where inheritance is useful in other ways, one should use `hasOwnProperty` instead. So as it stands, the above static rule would always be wrong for 50% of cases. I don't think we can or should ban use of object literals, nor should we ban use of `Object.create(null);`. These objects are different and both will be used.

Basically, we'd need to come up with a way to detect which is used, but I'm not sure that's possible so.. I guess this will just be a convention we can't statically detect/enforce.
